File: SixFour/Merge/MergeEvidence.swift
//  MergeEvidence.swift
//  SIGNAL FROM THE READS — the hand-written Swift twin of `Spec.MergeEvidence`.
//
//  The GIF is a gross approximation; the three reads are the evidence. So a
//  pour's deposit becomes a slot of an EVIDENCE SCHEDULE: 16 integers (one per
//  pour) derived DETERMINISTICALLY from the `S4TelemetrySnapshot` already
//  sealed beside the decision word in the same `.s4cr` record — replay stays a
//  pure function of `(schedule, word)` and the wire needs NO version bump
//  (`Spec.MergeEvidence.lawRecordedWordReplaysWithTelemetry`).
//
//  THE TOTAL RULE (`scheduleOf`): no snapshot → the derived constant;
//  otherwise the exact integer fair split of the burst's color-time budget —
//  priced by ARRIVALS ALONE, never the comovement byte (the 1000‰ sentinel
//  means both "derived" and "could not measure", so routing on it would let
//  an interrupted unmeasured burst masquerade as fully-funded:
//  `lawUnmeasuredCannotMasquerade`). Full bursts price past the window and
//  clamp to the constant by arithmetic (`lawFullBudgetYieldsConstant`).
//  THE COMPATIBILITY GUARANTEES (all pinned in the spec + mirrored in
//  `MergeEvidenceTests`):
//  • full-window budgets ⇒ today's constant 4s (`lawFullBudgetYieldsConstant`),
//  • a healthy full ladder (64/32/16 arrivals: 64·1+32·2+16·4 = 192 → clamp 64)
//    ⇒ the constant too (`lawHealthyLadderYieldsConstant`),
//  • the shipped weave plan (24/12/4: 24·1+12·2+4·4 = 64 = the FULL window)
//    ⇒ the constant — the one-sensor cadence mismatch absorbed lawfully
//    (`lawWeavePlanBudgetIsFullWindow`).
//  Honesty is a law, not a hope: short evidence cannot construct
//  (`lawShortEvidenceCannotConstruct` — the game refuses, never invents), the
//  measurement gate is orthogonal to evidence scaling
//  (`lawBankNeedsMeasurementUnderSchedule`), and the unlock stays monotone.
//
//  The step itself lives on `S4MergeBoard` (`step(_:schedule:)` — the state's
//  file owns the setters); this file owns the schedule DERIVATION. Pure
//  integer value math, zero dependencies (not even Foundation).

/// Schedule derivation — the twin of `Spec.MergeEvidence`'s pricing half.
enum S4MergeEvidence {

    /// The derived-mode schedule (re-exported from the board so both spellings
    /// name ONE array): `pourDeposit` (4) in every slot, sums to exactly
    /// `windowUnits` (64). Under it the game is byte-for-byte today's
    /// (`lawDerivedScheduleIsStep`).
    static var derivedSchedule: [Int] { S4MergeBoard.derivedSchedule }

    /// The 320 cs window in weave units (`Spec.WeaveOrder.windowUnits`) —
    /// DERIVED from the economy identity it prices
    /// (`Spec.MergeBoard.lawEconomyIsTheWindow`: window = threshold32 =
    /// `pourCap × pourDeposit`), never a free 64: if the economy ever moves,
    /// the budget ceiling moves with it by compile-time link.
    static let windowUnits = S4MergeBoard.pourCap * S4MergeBoard.pourDeposit

    /// Weave units one arrival of each rung spans, fine → coarse (the
    /// `S4TelemetrySnapshot.arrivals` order): `Spec.WeaveOrder.unitsOf` —
    /// DELEGATED to the one Swift owner of the pool-depth ladder
    /// (`ColorTimeDisplayMath.displayPeriodTicks` = 1/2/4), the same integers
    /// the display cadence and the slide's detents read.
    static let unitsPerRung = ColorTimeDisplayMath.displayPeriodTicks

    /// The burst's evidence budget in window units — the twin of
    /// `Spec.MergeEvidence.colorTimeBudget`: per rung, arrivals × `unitsOf`,
    /// summed and clamped to the window (a burst can never be worth more than
    /// its own 320 cs). Arrivals are wire-unsigned; each is pre-capped at
    /// `windowUnits` before the multiply — EXACT under the outer clamp
    /// (any arrival ≥ 64 already prices ≥ the whole window) and overflow-free
    /// for any `UInt64` the wire can carry. Missing entries read 0.
    static func colorTimeBudget(arrivals: [UInt64]) -> Int {
        var total = 0
        for (a, u) in zip(arrivals, unitsPerRung) {
            total += Int(min(a, UInt64(windowUnits))) * u
        }
        return min(windowUnits, total)
    }

    /// Exact integer fair division of a budget over the 16 pours — the twin of
    /// `Spec.MergeEvidence.fairSplit`: `e_i = ⌊(i+1)·a/16⌋ − ⌊i·a/16⌋`, which
    /// telescopes to exactly `a` and puts every slot at the floor or ceiling
    /// of the even share (`lawScheduleSplitConserves`). No floats, no
    /// remainder bias. Negative budgets clamp to 0; Int64 internals keep the
    /// multiply exact for any budget a caller can reach.
    static func fairSplit(_ budget: Int) -> [Int] {
        let a = Int64(max(0, min(budget, Int(Int32.max))))
        let cap = Int64(S4MergeBoard.pourCap)
        return (0 ..< S4MergeBoard.pourCap).map { i in
            Int((Int64(i + 1) * a) / cap - (Int64(i) * a) / cap)
        }
    }

    /// THE TOTAL RULE — the twin of `Spec.MergeEvidence.scheduleOf`: schedule
    /// from the (optional) sealed snapshot. No snapshot → `derivedSchedule`
    /// (v1 / v2-without-tel records replay today's game); otherwise
    /// `fairSplit` of the `colorTimeBudget` — priced by ARRIVALS ALONE,
    /// never the comovement byte (1000 permille means both "derived" and
    /// "could not measure", so routing on it would let an interrupted
    /// unmeasured burst masquerade as fully-funded —
    /// `lawUnmeasuredCannotMasquerade`). A full derived snapshot's arrivals
    /// price past the window and clamp to today's constant by arithmetic
    /// (`lawFullBudgetYieldsConstant`). A pure function of the snapshot — no
    /// floats, no clock, no hidden state, NO FEATURE FLAG: the live game and
    /// every future replay reader call this same rule on the same sealed
    /// bytes (`lawRecordedWordReplaysWithTelemetry`).
    static func schedule(from telemetry: S4TelemetrySnapshot?) -> [Int] {
        guard let telemetry else { return derivedSchedule }
        return fairSplit(colorTimeBudget(arrivals: telemetry.arrivals))
    }
}

File: spec/test/Properties/MergeEvidence.hs
module Properties.MergeEvidence (tests) where

import Test.Tasty
import Test.Tasty.QuickCheck

import SixFour.Spec.MergeEvidence
import SixFour.Spec.CaptureRecord (TelemetrySnapshot (..))
import SixFour.Spec.MergeBoard
  ( Board (..), GameOp (..), MoveOp (..), playAll, pourCap, regionCount )

-- Ops include out-of-range regions on purpose: totality is part of the
-- surface. Pours are common so random walks exercise the deposits.
genOp :: Gen GameOp
genOp = frequency
  [ (3, pure GPour)
  , (7, GMove <$> choose (-2, regionCount + 1)
              <*> elements [OpS, OpK, OpI])
  ]

genOps :: Gen [GameOp]
genOps = resize 120 (listOf genOp)

-- Hostile schedules: negative entries, wrong lengths (short AND long) — the
-- effectiveDeposit clamps are under test, not spared.
genSchedule :: Gen EvidenceSchedule
genSchedule = oneof
  [ vectorOf pourCap (choose (-8, 12))
  , resize 40 (listOf (choose (-8, 12)))
  , pure derivedSchedule
  , fairSplit <$> choose (0, 64)
  ]

genTelemetry :: Gen (Maybe TelemetrySnapshot)
genTelemetry = oneof
  [ pure Nothing
  , Just <$> (TelemetrySnapshot
                <$> resize 5 (listOf (choose (-4, 128)))
                <*> resize 5 (listOf (choose (0, 100000)))
                <*> choose (0, 1500))
  ]

tests :: TestTree
tests = testGroup "MergeEvidence (pour credit from the burst's OWN telemetry: derived special case, fair split, schedule-replayed keystone, wire-composed)"
  [ -- The derived special case -----------------------------------------------
    testProperty "lawDerivedScheduleIsStep: under the constant schedule playAllWith == playAll on EVERY op list" $
      forAll genOps lawDerivedScheduleIsStep

  , testProperty "lawFullBudgetYieldsConstant: a full-window budget (or no snapshot) earns exactly today's constant 4s, by arithmetic not signature" $
      \as com -> lawFullBudgetYieldsConstant as com

  , testProperty "lawUnmeasuredCannotMasquerade: short arrivals earn ONLY their own budget, whatever the comovement byte claims" $
      \a b c com -> lawUnmeasuredCannotMasquerade a b c com

  , testProperty "lawHealthyLadderYieldsConstant: arrivals 64/32/16 -> budget 192 -> clamp 64 -> constant 4" $
      once lawHealthyLadderYieldsConstant

  , testProperty "lawWeavePlanBudgetIsFullWindow: the shipped 24/12/4 weave prices at exactly the window (24+24+16=64)" $
      once lawWeavePlanBudgetIsFullWindow

    -- The fair split ---------------------------------------------------------
  , testProperty "lawScheduleSplitConserves: fairSplit sums to the budget, every slot floor-or-ceiling of the even share" $
      \a -> lawScheduleSplitConserves a

  , testProperty "goldenFairSplit: the pinned splits for budgets 0, 3, 24, 40, 64" $
      once (   fairSplit 0  == replicate 16 0
            && fairSplit 3  == [0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]
            && fairSplit 24 == [1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2]
            && fairSplit 40 == [2,3,2,3,2,3,2,3,2,3,2,3,2,3,2,3]
            && fairSplit 64 == replicate 16 4 )

  , testProperty "lawScheduleNeverExceedsWindow: scheduleOf is 16 non-negative slots summing to at most the window" $
      forAll genTelemetry lawScheduleNeverExceedsWindow

    -- The keystones ----------------------------------------------------------
  , testProperty "lawWordReplaysBoardUnderSchedule (GENERALIZED KEYSTONE): the word replays the board under ITS schedule" $
      forAll ((,) <$> genSchedule <*> genOps) (uncurry lawWordReplaysBoardUnderSchedule)

  , testProperty "lawRecordedWordReplaysWithTelemetry (WIRE KEYSTONE): seal tel+dw in ONE v3 record, decode, replay — same board" $
      \as ns como -> forAll genOps $ \ops ->
        lawRecordedWordReplaysWithTelemetry as ns como ops

    -- Conservation and honesty -----------------------------------------------
  , testProperty "lawSignalLedgerConservedUnderSchedule: signal == effective deposits - spends, never negative, pours capped" $
      forAll ((,) <$> genSchedule <*> genOps) (uncurry lawSignalLedgerConservedUnderSchedule)

  , testProperty "lawBankNeedsMeasurementUnderSchedule: all-coarse pours bank ZERO under ANY schedule (the gate is orthogonal)" $
      forAll ((,) <$> genSchedule <*> choose (0, 32)) (uncurry lawBankNeedsMeasurementUnderSchedule)

  , testProperty "lawUnlockMonotoneUnderSchedule: banked 32-evidence never decreases under any schedule" $
      forAll ((,) <$> genSchedule <*> genOps) (uncurry lawUnlockMonotoneUnderSchedule)

  , testProperty "lawShortEvidenceCannotConstruct: total evidence < 48 packets -> no op list fully constructs (refuse, never invent)" $
      forAll ((,) <$> genSchedule <*> genOps) (uncurry lawShortEvidenceCannotConstruct)

    -- The pinned evidence run ------------------------------------------------
  , testProperty "golden evidence trace: canonicalEvidenceRun under fairSplit 40 — every field pinned, refusal absent, replay exact" $
      once $
        let s = fairSplit 40
            b = playAllWith s canonicalEvidenceRun
        in bDepths b == ([0, 1, 1] ++ replicate 13 0)
           && bSignal b == 7
           && bSpent b == 3
           && bPours b == 4
           && bBank32 b == 10
           && bWord b == [ GPour, GMove 0 OpS, GPour, GMove 1 OpS, GPour
                         , GMove 0 OpK, GPour, GMove 2 OpS, GMove 3 OpI ]
           && playAllWith s (bWord b) == b
           && playAll canonicalEvidenceRun /= b   -- the constant game DIFFERS: evidence scaling is real
  ]
